designs/2024-baseline-support/README.md

```diff
@@ -0,0 +1,371 @@
+   * If no, keep the error.
+3. **Report and exit**
+    * Exit with a non-zero status if there are unmatched violations, optionally listing them in verbose mode.
+    * Otherwise, list remaining errors as usual.
+
+Note that the error detection in `cli.js` occurs before the error counting. This allow us to update the suppressions file and modify the errors, before it is time to count errors. Please refer to the last example of the "Implementation notes" for more details.
+
+Furthermore, ESLint cache (`--cache`) must include the full list of detected violations, even those in the suppressions file. This approach has the following benefits:
+
+- Generating the suppressions file can be based on the cache file and should be faster when the cache file is used.
+- Allows developers to update the suppressions file and then re-lint still taking the same cache into consideration.
+- It even allows developers to delete the suppressions file and still take advantage of the cached file in subsequent runs. 
+
+### Implementation notes
+
+To introduce the above-mentioned options, we will need to: 
+
+* add the new options in `default-cli-options.js`.
+* adjust the config for optionator.
+* add the new options as comments and arguments for eslint.
+* update documentation to explain the newly introduced feature.
+
+A new type must be created to represent the suppressions file:
+
+``` js
+/**
+ * @typedef {Record<string, Record<string, { count: number }>} SuppressedViolations
+ */
+```
+
+A new class must be created to manage the suppressions file:
+
+``` js
+class SuppressedViolationsManager {
+    /**
+     * Creates a new instance of SuppressedViolationsManager.
+     * @param {string} suppressionsLocation The location of the suppressions file.
+     */
+    constructor(suppressionsLocation) {}
+
+    /**
+     * Updates the suppressions file based on the current violations.
+     * 
+     * @param {LintResult[]} results The lint results.
+     * @returns {void}
+     */
+    suppressAll(results) 
+
+    /**
+     * Updates the suppressions file based on the current violations and the provided rule.
+     * 
+     * @param {LintResult[]} results The lint results.
+     * @param {string[]} rules The rules to suppress.
+     * @returns {void}
+     */
+    suppressRules(results, rules)
+
+    /**
+     * Removes old suppressions that do not occur anymore.
+     * @returns {void}
+     */
+    prune()
+
+    /**
+     * Checks the provided suppressions against the lint results.
+     * It filters and returns the lint results that are not in the suppressions file.
+     * It also returns the unmatched suppressions.
+     * 
+     * @param {LintResult[]} results The lint results.
+     * @param {SuppressedViolations} suppressions The suppressions.
+     * @returns {{
+     *   results: LintResult[],
+     *   unmatched: SuppressedViolations
+     * }}
+     */
+    applySuppressions(results, suppressions)
```

Comment:
The suppressions file only knows how many results should be suppressed, not which ones. How will the filtering work exactly?

Reply:
It was described in "Match Errors Against Suppressions". I have updated the documentation here to reflect the same flow.